File: leetcode/123/MaxProfit.java
import java.util.Arrays;

import javax.sound.midi.Soundbank;

class MaxProfit {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.maxProfit(new int[] { 3, 3, 5, 0, 0, 3, 1, 4 }));

        Solution1 s1 = new Solution1();
        System.out.println(Arrays.toString(new int[] { 3, 3, 5, 0, 0, 3, 1, 4 }));
        System.out.println(s1.maxProfit(new int[] {3,3,5,0,0,3,1,4}));

        Solution2 s2 = new Solution2();
        System.out.println(s2.maxProfit(new int[] { 3, 3, 5, 0, 0, 3, 1, 4 }));
    }
}
    
class Solution {
    public int maxProfit(int[] prices) {
        int buy1 = Integer.MIN_VALUE;
        int buy2 = Integer.MIN_VALUE;
        int sell1 = 0;
        int sell2 = 0;

        for (int i = 0; i < prices.length; i++) {
            buy1 = Math.max(buy1, -prices[i]);
            sell1 = Math.max(sell1, buy1 + prices[i]);
            buy2 = Math.max(buy2, sell1 - prices[i]);
            sell2 = Math.max(sell2, buy2 + prices[i]);
        }
        return Math.max(sell1, sell2);
    }
}

class Solution1{
    public int maxProfit(int[] prices) {
        int firstBuy = Integer.MAX_VALUE;
        int secondBuy = Integer.MAX_VALUE;
        int firstSell = 0;
        int secondSell = 0;
        
        System.out.println("firstBuy" + "," + "firstSell" + "," + "secondBuy" + "," + "secondSell");    
        for(int i = 0; i < prices.length; i++)
        {
            firstBuy = Math.min(firstBuy,prices[i]);
            firstSell = Math.max(firstSell,prices[i] - firstBuy);
            secondBuy = Math.min(secondBuy,prices[i] - firstSell);
            secondSell = Math.max(secondSell, prices[i] - secondBuy);
    
    
            System.out.println(firstBuy+"       ,"+firstSell+"        ,"+secondBuy+"        ,"+secondSell);
        } 
    // 
        return secondSell;
    }
}

class Solution2 {
    public int maxProfit(int[] prices) {
        if (prices.length < 2)
            return 0;

        int buy1 = prices[0];
        int sell1 = 0;

        int buy2 = prices[0];
        int sell2 = 0;

        for (int i = 0; i < prices.length; i++) {

            if (buy1 > prices[i])
                buy1 = prices[i];

            if (sell1 < prices[i] - buy1)
                sell1 = prices[i] - buy1;

            if (buy2 > prices[i] - sell1)
                buy2 = prices[i] - sell1;
            if (sell2 < prices[i] - buy2)
                sell2 = prices[i] - buy2;

        }
        return sell2;
    }
}
                     

    

            

                         

    

            

                         

    

            

                         

    

            

                           

                     

    

            

                         

    

            

        

    

            

                         

    

            

                         

    

            

                 
       
       
       
       
       

     
       
       
       
       
       
       
        
       
       

    
       

    
     

    
    
             

          
          
  
         

        

File: leetcode/123/MaxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
        let firstBuy = Number.MAX_VALUE;
        let secondBuy = Number.MAX_VALUE;
        let firstSell = 0;
        let secondSell = 0;

        for(let i = 0; i < prices.length; i++)
        {
            firstBuy = Math.min(firstBuy,prices[i]);
            firstSell = Math.max(firstSell,prices[i] - firstBuy);
            secondBuy = Math.min(secondBuy,prices[i] - firstSell);
            secondSell = Math.max(secondSell, prices[i] - secondBuy);

            // console.log(firstBuy+","+prices[i]);
            // console.log(firstSell+","+(prices[i] - firstBuy));
            // console.log(secondBuy+","+(prices[i] - firstSell));
            // console.log(secondSell+","+(prices[i] - secondBuy));

            // console.log("prices["+i+"]," + "firstBuy"+","+"firstSell"+","+"secondBuy"+","+"secondSell");
            // console.log(prices[i] +"         " + firstBuy+",        "+firstSell+",       "+secondBuy+",        "+secondSell);

            // console.log("====================================");
        } 
    // 
        return secondSell;
};

var maxProfit1 = function(prices) {
        let firstBuy = -Number.MAX_VALUE;
        let secondBuy = -Number.MAX_VALUE;
        let firstSell = 0;
        let secondSell = 0;

        for(let i = 0; i < prices.length; i++)
        {
            firstBuy = Math.max(firstBuy, -prices[i]);
            firstSell = Math.max(firstSell, firstBuy + prices[i]) ;
            secondBuy = Math.max(secondBuy, firstSell - prices[i]);
            secondSell = Math.max(secondSell, secondBuy + prices[i]);

            // console.log(firstBuy+","+prices[i]);
            // console.log(firstSell+","+(prices[i] - firstBuy));
            // console.log(secondBuy+","+(prices[i] - firstSell));
            // console.log(secondSell+","+(prices[i] - secondBuy));

            // console.log("prices["+i+"]," + "firstBuy"+","+"firstSell"+","+"secondBuy"+","+"secondSell");
            // console.log(prices[i] +"         " + firstBuy+",        "+firstSell+",       "+secondBuy+",        "+secondSell);
            console.log(firstBuy +"," + firstSell+","+secondBuy+","+secondSell);
            // console.log("====================================");
        } 
    // 
        return Math.max(secondSell);
};

var maxProfit2 = function(prices) {
        // let fb = Number.NEGATIVE_INFINITY;
        // let sb = Number.NEGATIVE_INFINITY;
        let fb = -Number.MAX_VALUE;
        let sb = -Number.MAX_VALUE;
        let fs = 0;
        let ss = 0;


        console.log("Number.MIN_VALUE : " + Number.NEGATIVE_INFINITY);

        for(let i = 0; i < prices.length; i++)
        {
            fb = Math.max(fb, -prices[i]);
            fs = Math.max(fs, fb + prices[i]) ;
            sb = Math.max(sb, fs - prices[i]);
            ss = Math.max(ss, sb + prices[i]);
            console.log(fb +"," + fs+","+sb+","+ss);
        } 
    // 
        return ss;
};


console.log(maxProfit1([3, 3, 5, 0, 0, 3, 1, 4 ]));
console.log(maxProfit2([3, 3, 5, 0, 0, 3, 1, 4 ]));
/*
prices[0],firstBuy,firstSell,secondBuy,secondSell
3         3,        0,       3,        0
====================================
prices[1],firstBuy,firstSell,secondBuy,secondSell
3         3,        0,       3,        0
====================================
prices[2],firstBuy,firstSell,secondBuy,secondSell
5         3,        2,       3,        2
====================================
prices[3],firstBuy,firstSell,secondBuy,secondSell
0         0,        2,       -2,        2
====================================
prices[4],firstBuy,firstSell,secondBuy,secondSell
0         0,        2,       -2,        2
====================================
prices[5],firstBuy,firstSell,secondBuy,secondSell
3         0,        3,       -2,        5
====================================
prices[6],firstBuy,firstSell,secondBuy,secondSell
1         0,        3,       -2,        5
====================================
prices[7],firstBuy,firstSell,secondBuy,secondSell
4         0,        4,       -2,        6
====================================
*/